datafusion/physical-plan/src/joins/hash_join.rs

```diff
@@ -1064,149 +1149,234 @@ pub fn equal_rows_arr(
 
 impl HashJoinStream {
     /// Separate implementation function that unpins the [`HashJoinStream`] so
-    /// that partial borrows work correctly
+    /// that partial borrows work correctly.
+    ///
+    /// Performs state transitions for HashJoinStream according to following diagram:
```

Comment:
This state diagram might be easier to find if it were on `HashJoinStreamState` itself